### PR TITLE
Add a config to allow user config main container resource

### PR DESCRIPTION
**Pull request by ywskycn**
Adding the MainContainerResources config, to make executeResource can work if resource is required. We may export the config also for executeContainer and executeScript funcs.

### CODE REVIEW

**Comment by xianlubird** [comment]
Maybe we can config this is template yaml rather than in the common config

**Comment by ywskycn** [comment]
> Maybe we can config this is template yaml rather than in the common config

you mean in the workflow yaml?

**Comment by ywskycn** [comment]
@jessesuen mind take a look?

File: workflow/controller/operator.go [hunk 1]
```diff
@@ -1487,6 +1487,9 @@ func (woc *wfOperationCtx) executeResource(nodeName string, tmpl *wfv1.Template,
 		},
 		Env: execEnvVars,
 	}
+	if woc.controller.Config.MainContainerResources != nil {
+		mainCtr.Resources = *woc.controller.Config.MainContainerResources
```
Comment (jessesuen) [inline workflow/controller/operator.go]: It seems this logic will blow away the resources that the user defines. We need to honor the resources defined in the main container, and only fall back to the controller config if unspecified.

Comment (ywskycn) [reply workflow/controller/operator.go]: but in executeResource, users don't have the opportunity to specify the resource, right? As the mainCtr is created by argo itself. Or I miss anything here?

Comment (ywskycn) [reply workflow/controller/operator.go]: @jessesuen   comments?

**Comment by alexec** [comment]
I think this has been superseded by defaultWorkflowSpec?

**Comment by ywskycn** [comment]
> I think this has been superseded by defaultWorkflowSpec?

#2331 ? I'm not sure how we config the main container resource for resource template through the defaultWorkflowSpec here.

**Comment by alexec** [comment]
No activity in 6 months - closing.

**Pull request closed without merging**